server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -21,79 +21,13 @@
-import java.security.PrivilegedAction;
-import java.util.BitSet;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.LinkedHashMap;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
-import java.util.concurrent.CopyOnWriteArraySet;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 /**
  * A logger that logs deprecation notices.
  */
 public class DeprecationLogger {
-
-    private final Logger logger;
-
-    /**
-     * This is set once by the {@code Node} constructor, but it uses {@link CopyOnWriteArraySet} to ensure that tests can run in parallel.
-     * <p>
-     * Integration tests will create separate nodes within the same classloader, thus leading to a shared, {@code static} state.
-     * In order for all tests to appropriately be handled, this must be able to remember <em>all</em> {@link ThreadContext}s that it is
-     * given in a thread safe manner.
-     * <p>
-     * For actual usage, multiple nodes do not share the same JVM and therefore this will only be set once in practice.
-     */
-    private static final CopyOnWriteArraySet<ThreadContext> THREAD_CONTEXT = new CopyOnWriteArraySet<>();
-
-    /**
-     * Set the {@link ThreadContext} used to add deprecation headers to network responses.
-     * <p>
-     * This is expected to <em>only</em> be invoked by the {@code Node}'s constructor (therefore once outside of tests).
-     *
-     * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
-     * @throws IllegalStateException if this {@code threadContext} has already been set
-     */
-    public static void setThreadContext(ThreadContext threadContext) {
-        Objects.requireNonNull(threadContext, "Cannot register a null ThreadContext");
-
-        // add returning false means it _did_ have it already
-        if (THREAD_CONTEXT.add(threadContext) == false) {
-            throw new IllegalStateException("Double-setting ThreadContext not allowed!");
-        }
-    }
-
-    /**
-     * Remove the {@link ThreadContext} used to add deprecation headers to network responses.
-     * <p>
-     * This is expected to <em>only</em> be invoked by the {@code Node}'s {@code close} method (therefore once outside of tests).
-     *
-     * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
-     * @throws IllegalStateException if this {@code threadContext} is unknown (and presumably already unset before)
-     */
-    public static void removeThreadContext(ThreadContext threadContext) {
-        assert threadContext != null;
-
-        // remove returning false means it did not have it already
-        if (THREAD_CONTEXT.remove(threadContext) == false) {
-            throw new IllegalStateException("Removing unknown ThreadContext not allowed!");
-        }
-    }
+    private final ThrottlingAndHeaderWarningLogger deprecationLogger;
```

Comment:
Can you beef up the class level doc a little bit ?   Just a quick mention that it does a throttle log (and what it throttling ..or link to the throttling class) and a warning message. Also a small usage in the description would be great.

Reply:
good idea, added usage and more detailed javadoc.